src/grammar.coffee

```diff
@@ -468,6 +469,112 @@ grammar =
+  ]
+
+  JsxElementShorthands: [
+    o 'JsxElementShorthandClasses',                                                  -> classes: $1
+    o 'JsxElementShorthandId',                                                       -> id: $1
+    o 'JsxElementShorthandId JsxElementShorthandClasses',                            -> classes: $2, id: $1
+    o 'JsxElementShorthandClasses JsxElementShorthandId JsxElementShorthandClasses', -> classes: $1.concat($2), id: $2
+    o 'JsxElementShorthandClasses JsxElementShorthandId',                            -> classes: $1, id: $2
+  ]
+
+  JsxElementShorthandId: [
+    o 'JSX_ID_SHORTHAND_SYMBOL JSX_ID_SHORTHAND', -> $2
+  ]
+
+  JsxElementShorthandClasses: [
+    o 'JsxElementShorthandClass',                            -> [$1]
+    o 'JsxElementShorthandClasses JsxElementShorthandClass', -> $1.concat $2
+  ]
+
+  JsxElementShorthandClass: [
+    o 'JSX_CLASS_SHORTHAND_SYMBOL JSX_CLASS_SHORTHAND', -> $2
+  ]
+
+  JsxAttributes: [
+    o 'JsxParenthesizedAttributes_',                      -> list: $1
+    o 'JsxParenthesizedAttributes_ JsxObjectAttributes_', -> list: $1, object: $2
+    o 'JsxObjectAttributes_ JsxParenthesizedAttributes_', -> list: $2, object: $1
+    o 'JsxObjectAttributes_',                             -> object: $1
+  ]
+
+  JsxParenthesizedAttributes_: [
+    o 'JSX_PARENTHESIZED_ATTRIBUTES_START JsxParenthesizedAttributes JSX_PARENTHESIZED_ATTRIBUTES_END', -> $2
+  ]
+
+  JsxParenthesizedAttributes: [
+    o 'JsxParenthesizedAttribute',                            -> [$1]
+    o 'JsxParenthesizedAttributes JsxParenthesizedAttribute', -> $1.concat $2
+    o 'INDENT JsxParenthesizedAttributes OUTDENT',            -> $2
+    o 'JsxParenthesizedAttributes TERMINATOR',                -> $1
+    o 'JsxParenthesizedAttributes INDENT JsxParenthesizedAttributes OUTDENT', -> $1.concat $3
+  ]
+
+  JsxParenthesizedAttribute: [
+    o 'JSX_ATTRIBUTE_NAME = JsxParenthesizedAttributeValue', -> name: $1, value: $3
+  ]
+
+  JsxParenthesizedAttributeValue: [
+    o 'STRING',         -> new StringLiteral $1
+    o '{ Expression }', -> $2
+  ]
+
+  JsxObjectAttributes_: [
+    o 'JSX_OBJECT_ATTRIBUTES_START JsxAttributesObject JSX_OBJECT_ATTRIBUTES_END', -> $2
+  ]
+
+  JsxAttributesObject: [
+    o '{ AssignList OptComma }', -> new JsxAttributesObj $2, $1.generated
+  ]
+
+  JsxElementChildren_: [
+    o 'JsxElementChildren JSX_ELEMENT_INLINE_BODY_END', -> $1
+    o 'INDENT JsxElementChildren OUTDENT',              -> $2
+  ]
+
+  JsxElementChildren: [
+    o 'JsxElementChild',                    -> [$1]
+    o 'JsxElementChildren JsxElementChild', -> $1.concat $2
+  ]
+
+  JsxElementChild: [
+    o 'JSX_ELEMENT_CONTENT',           -> $1
+    o 'JsxElement',                    -> $1
+    o '{ Expression }',                -> $2
+    o '{ INDENT Expression OUTDENT }', -> $3
+  ]
```

Comment:
@GeoffreyBooth I hear you, that's part of what made me uneasy about whether this belongs in Coffeescript proper. I took a closer look at @xixixao's approach in #4551 and it's super elegant/lightweight. But it relies on having a known end tag to be able to treat JSX element bodies like specially-interpolated strings. If you want whitespace-indented tags I think you have to go more heavyweight as far as lexing differently while you're inside an element body and then having grammar rules (at least to the extent of representing the indentation structure). I do think there's a demand for the whitespace-indented version of JSX, it's certainly what I wanted (thus the itch-scratching) and I saw your comments wishing for a Jade-like syntax (which is pretty similar to what I've implemented). But I'm inclined to think that @xixixao's version, being less breaking-changes-y, much more lightweight, and just actual JSX syntactically (with interpolated Coffeescript) is what belongs in Coffeescript proper. And then I'm not sure exactly what constitutes maintaining something like this (releasing a separate NPM package? periodically merging in `master`/`2`?) but I'd think that I could maintain this syntactically sweeter version separately and try to make it visible for those who might prefer it, what do you think?